openshift-knative-operator/pkg/monitoring/monitoring.go

```diff
@@ -21,45 +22,76 @@ import (
+	EnableEventingMonitoringEnvVar = "ENABLE_EVENTING_MONITORING_BY_DEFAULT"
+	EnableMonitoringLabel          = "openshift.io/cluster-monitoring"
+	ObservabilityCMName            = "observability"
+	ObservabilityBackendKey        = "metrics.backend-destination"
+	OpenshiftMonitoringNamespace   = "openshift-monitoring"
+	prometheusRoleName             = "knative-prometheus-k8s"
+	prometheusClusterRoleName      = "rbac-proxy-metrics-prom"
+	smRbacManifestPath             = "SERVICE_MONITOR_RBAC_MANIFEST_PATH"
+)
+
+type KComponentType string
+
+const (
+	Serving  KComponentType = "Serving"
+	Eventing KComponentType = "Eventing"
 )
 
 var (
-	servingComponents = sets.NewString("activator", "autoscaler", "autoscaler-hpa", "controller", "domain-mapping", "domainmapping-webhook", "webhook")
+	servingComponents  = sets.NewString("activator", "autoscaler", "autoscaler-hpa", "controller", "domain-mapping", "domainmapping-webhook", "webhook")
+	eventingComponents = sets.NewString("eventing-controller", "eventing-webhook", "imc-controller", "imc-dispatcher", "mt-broker-controller", "mt-broker-filter", "mt-broker-ingress", "sugar-controller")
 )
 
 func init() {
 	_ = monitoringv1.AddToScheme(scheme.Scheme)
 }
 
-func ReconcileServingMonitoring(ctx context.Context, api kubernetes.Interface, ks *v1alpha1.KnativeServing) error {
-	backend := ks.Spec.CommonSpec.Config[ObservabilityCMName][ObservabilityBackendKey]
-	if shouldEnableMonitoring(backend) {
-		if err := reconcileMonitoringLabelOnNamespace(ctx, ks.Namespace, api, true); err != nil {
+func ReconcileMonitoringForKComponent(ctx context.Context, api kubernetes.Interface, comp v1alpha1.KComponent) error {
+	commonSpec, cType := getCommonSpec(comp)
+	if shouldEnableMonitoring(commonSpec, cType) {
+		if err := reconcileMonitoringLabelOnNamespace(ctx, comp.GetNamespace(), api, true, cType); err != nil {
 			return fmt.Errorf("failed to enable monitoring %w ", err)
 		}
 		return nil
 	}
-	if err := reconcileMonitoringLabelOnNamespace(ctx, ks.Namespace, api, false); err != nil {
+	// If "opencensus" is used we still dont want to scrape from a Serverless controlled namespace
+	// user can always push to an agent collector in some other namespace and then integrate with OCP monitoring stack
+	if err := reconcileMonitoringLabelOnNamespace(ctx, comp.GetNamespace(), api, false, cType); err != nil {
 		return fmt.Errorf("failed to disable monitoring %w ", err)
 	}
-	common.Configure(&ks.Spec.CommonSpec, ObservabilityCMName, ObservabilityBackendKey, "none")
+	common.Configure(commonSpec, ObservabilityCMName, ObservabilityBackendKey, "none")
 	return nil
 }
 
-func shouldEnableMonitoring(backend string) bool {
-	if backend == "none" {
+func getCommonSpec(comp v1alpha1.KComponent) (*v1alpha1.CommonSpec, KComponentType) {
+	switch c := comp.(type) {
+	case *v1alpha1.KnativeServing:
+		return &c.Spec.CommonSpec, Serving
+	case *v1alpha1.KnativeEventing:
+		return &c.Spec.CommonSpec, Eventing
+	}
+	return nil, ""
+}
```

Comment:
Is this necessary? Why can't we use `comp.GetSpec()`?

Reply:
Because we want to the common Spec not the direct Spec for each comp. Common Spec is embedded in the top level spec.